meja/tests/row-types.meja

```diff
@@ -0,0 +1,368 @@
+module type Fixed_rows = {
+  let no_arg : [`A];
+
+  let no_args : [`A | `B];
+
+  let arg : [`A(int)];
+
+  let args : [`A(int) | `B(bool)];
+
+  let arg_noarg : [`A(int) | `B];
+
+  let noarg_arg : [`A | `B(bool)];
+};
+
+module type Open_rows = {
+  let empty : [>];
+
+  let no_arg : [> `A];
+
+  let no_args : [>`A | `B];
+
+  let arg : [>`A(int)];
+
+  let args : [>`A(int) | `B(bool)];
+
+  let arg_noarg : [>`A(int) | `B];
+
+  let noarg_arg : [>`A | `B(bool)];
+};
+
+module type Closed_rows = {
+  let no_arg : [< `A];
+
+  let no_args : [<`A | `B];
+
+  let arg : [<`A(int)];
+
+  let args : [<`A(int) | `B(bool)];
+
+  let arg_noarg : [<`A(int) | `B];
+
+  let noarg_arg : [<`A | `B(bool)];
+};
+
+module type Bounded_rows = {
+  let fixed_no_args : [< `A | `B > `A | `B ];
```

Comment:
What is the meaning of this kind of type? Do you think you could put a spec somewhere? 

Also - do you think you could get requests working in an untyped way (like we have with the library) to get to a working version sooner or is it somehow entangled?

Reply:
Somehow managed to miss this review, sorry.

This type is analogous to the OCaml type of the same form:
```ocaml
[< `Tag_1 | `Tag_2 | `Tag_3 | ... | `Tag_n > `Tag_i | `Tag_j | ... | `Tag_k]
```
is a type that may only be constructed by the constructors to the left of `<`, and the minimum set of constructors are those to the right of `>`. For example,
```ocaml
val f : [< `A | `B | `C > `A] -> _

(* This typechecks, restricting the type to the minimum set of constructors. *)
let g (x : [`A]) = f x

(* This typechecks, all of the constructors are present. *)
let h (x : [`A | `B | `C]) = f x
```



> Also - do you think you could get requests working in an untyped way (like we have with the library) to get to a working version sooner or is it somehow entangled?

Requests should currently work, but we have to attach handlers manually with `handle`, as in OCaml. The row typing work is to infer a) when we have a handler that we can attach, so that we can do so manually and b) when we try to compile a whole program but there is no handler for a request that isn't hit.